### PR TITLE
Addition of Microsoft Cognitive Translation Service and a Tab to select the preferred Translations service

### PR DESCRIPTION
**Pull request by SudarshanSMD**
### Addition of Microsoft Cognitive Translations service
We have leveraged Microsoft's Cognitive Translations service for providing the translations. We already have free translations service provided by Google. However, the free service by Google has limited throughput. This is noticeable especially when translation resources file.
Addition of Microsoft's Cognitive Translations service would enable people who have a subscription to the Azure Cognitive services to leverage it and get fast performance.

### Addition of new tab for selecting the preferred Translations service
We have added a new tab for a selection of Preferred Translation service. By default, Google translation service is used. User can select the Microsoft radio button to change the preferred translation service to Microsoft Cognitive Translations service.
frmMain has the PreferredService variable, depending on which respective service is invokes for translation.

### CODE REVIEW

**Comment by salarcode** [comment]
Thanks, I'll review it later on

**Comment by SudarshanSMD** [comment]
I've added a new commit to fix a bug in MSTranslationService.

> We had a bug in MSTranslations service. If the translations were made wherein the source language was explicitly specified, we would get an exception. It was just a just matter of no null check for the detected language. We have removed the line for getting detected language, as it was just used for logging on the console.

**Comment by salarcode** [comment]
There is a lot of bad design decisions you have made. I'll merge this request and fix them in the next commit. Good learning point for you.
Also i'll mention them as a comment here.

File: AutoResxTranslator/AutoResxTranslator.csproj [hunk 1]
```diff
@@ -59,11 +59,13 @@
       <DependentUpon>frmMain.cs</DependentUpon>
     </Compile>
     <Compile Include="GTranslateService.cs" />
+    <Compile Include="MSTranslateService.cs" />
     <Compile Include="OpResult.cs" />
     <Compile Include="Program.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
     <Compile Include="ResxExcel.cs" />
     <Compile Include="ResxTranslator.cs" />
+    <Compile Include="ServiceEnum.cs" />
```
Comment (salarcode) [inline AutoResxTranslator/AutoResxTranslator.csproj]: Definitions should go into a dedicated folder

File: AutoResxTranslator/MSTranslateService.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+namespace AutoResxTranslator
+{
+    /// <summary>
+    /// Translation service using Microsoft Cogntive service.
+    /// 
+    /// ref: https://azure.microsoft.com/en-in/services/cognitive-services/translator-text-api/
+    /// </summary>
+    public class MSTranslateService
+    {
+        private static string host = "https://api-apc.cognitive.microsofttranslator.com";
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: this should be const

File: AutoResxTranslator/MSTranslateService.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        private static string subscriptionKey = "<YOUR-SUBSCRIPTION-KEY-HERE>";
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: this should be a input from UI

File: AutoResxTranslator/MSTranslateService.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                fromLanguage = null;
+            }
+
+            string route = "/translate?api-version=3.0&to=" + toLanguage + "&from=" + fromLanguage;
+
+            try
+            {
+                object[] body = new object[] { new { Text = text } };
+                var requestBody = JsonConvert.SerializeObject(body);
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: Where is the package reference to JSON.NET in the project?

File: AutoResxTranslator/MSTranslateService.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+                using (var client = new HttpClient())
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: The project uses .net 4 you should have updated to v4.5 in the project to use HttpClient

File: AutoResxTranslator/MSTranslateService.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                            // Iterate over the results, return the first result
+                            foreach (Translation t in o.Translations)
+                            {
+                                return t.Text;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        return "Translation failed! Exception: " + response.ReasonPhrase;
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: Unknown status of the translation. The result should consist of a success flag and the value

File: AutoResxTranslator/MSTranslateService.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                return "Translation failed! Exception: " + e.Message;
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: same as above

File: AutoResxTranslator/MSTranslateService.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            }
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// The C# classes that represents the JSON returned by the Translator Text API.
+    /// </summary>
+    public class TranslationResult
```
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: dedicated folder!

File: AutoResxTranslator/ServiceEnum.cs [hunk 9]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AutoResxTranslator
+{
+    public enum ServiceEnum
```
Comment (salarcode) [inline AutoResxTranslator/ServiceEnum.cs]: Dedicated folder

File: AutoResxTranslator/frmMain.Designer.cs [hunk 10]
```diff
@@ -28,620 +28,676 @@ protected override void Dispose(bool disposing)
...
+            this.tabPage3.Controls.Add(this.groupBox2);
+            this.tabPage3.Location = new System.Drawing.Point(4, 26);
+            this.tabPage3.Name = "tabPage3";
+            this.tabPage3.Padding = new System.Windows.Forms.Padding(3);
+            this.tabPage3.Size = new System.Drawing.Size(642, 342);
+            this.tabPage3.TabIndex = 5;
+            this.tabPage3.Text = "Translate Service";
+            this.tabPage3.UseVisualStyleBackColor = true;
+            // 
+            // groupBox2
```
Comment (salarcode) [inline AutoResxTranslator/frmMain.Designer.cs]: groupBox2 should have anchor

File: AutoResxTranslator/frmMain.cs [hunk 11]
```diff
@@ -19,700 +19,737 @@
...
-			if (doc != null)
-				foreach (HtmlElement imgElemt in doc.Images)
-				{
-					imgElemt.SetAttribute("src", "");
-				}
-		}
-	}
+    public partial class frmMain : Form
+    {
+        public static ServiceEnum PreferedService = ServiceEnum.Google;
```
Comment (salarcode) [inline AutoResxTranslator/frmMain.cs]: Very bad design decision, this should be passed as parameter

File: AutoResxTranslator/frmMain.cs [hunk 12]
```diff
@@ -19,700 +19,737 @@
...
+            if (doc != null)
+                foreach (HtmlElement imgElemt in doc.Images)
+                {
+                    imgElemt.SetAttribute("src", "");
+                }
+        }
+
+        private void radioButtonGoogleTranslateService_CheckedChanged(object sender, EventArgs e)
+        {
+            if (sender is RadioButton rb && rb.Checked)
```
Comment (salarcode) [inline AutoResxTranslator/frmMain.cs]: if (radioButtonGoogleTranslateService.Checked) was enough

File: AutoResxTranslator/MSTranslateService.cs (hunk 8 above)
Comment (salarcode) [inline AutoResxTranslator/MSTranslateService.cs]: These shold've had a dedicated folder

**Comment by salarcode** [comment]
I've fixed your merge, check it out here:
https://github.com/salarcode/AutoResxTranslator/commit/8fdcad31ced167f53020c6aa5b3d4dae67537e0f

**Comment by SudarshanSMD** [comment]
@salarcode Thank you for the review and the suggestions on the better approach.


I should have spent some more time on the code.

File: AutoResxTranslator/frmMain.Designer.cs (hunk 10 above)
Comment (SudarshanSMD) [reply AutoResxTranslator/frmMain.Designer.cs]: Haven't worked much on WinForms, so I still would have missed this.
Thank you.

**Pull request merged**